File: java/src/main/java/com/zipcode/exercises/task05/MapDictionaryExercises.java
package com.zipcode.exercises.task05;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * Task 5: Maps and Dictionaries - Java
 * 
 * This class contains exercises for learning Java's Map interface and its implementations.
 * Maps are fundamental data structures that store key-value pairs, enabling efficient
 * lookups, storage, and manipulation of associative data.
 * 
 * Key concepts covered:
 * - HashMap, LinkedHashMap, TreeMap usage
 * - Key-value pair operations
 * - Map iteration patterns
 * - Frequency counting and grouping
 * - Map-based algorithms
 * - Performance considerations
 * 
 * Each method should be implemented to pass the corresponding unit tests.
 * Focus on understanding different Map implementations and their use cases.
 */
public class MapDictionaryExercises {

    /**
     * Create a map from arrays of keys and values.
     * 
     * Example:
     * createMap(["a", "b", "c"], [1, 2, 3]) → {"a": 1, "b": 2, "c": 3}
     * 
     * @param keys Array of keys
     * @param values Array of values
     * @return Map with paired keys and values
     */
    public Map<String, Integer> createMap(String[] keys, Integer[] values) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Get a value from the map, returning a default if key doesn't exist.
     * 
     * Example:
     * getValueOrDefault({"a": 1, "b": 2}, "a", 0) → 1
     * getValueOrDefault({"a": 1, "b": 2}, "c", 0) → 0
     * 
     * @param map The map to search
     * @param key The key to look up
     * @param defaultValue Value to return if key not found
     * @return The value for the key, or default value
     */
    public Integer getValueOrDefault(Map<String, Integer> map, String key, Integer defaultValue) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Count the frequency of each character in a string.
     * 
     * Example:
     * countCharacterFrequency("hello") → {'h': 1, 'e': 1, 'l': 2, 'o': 1}
     * 
     * @param text Input string
     * @return Map of character frequencies
     */
    public Map<Character, Integer> countCharacterFrequency(String text) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Find the most frequent element in an array.
     * 
     * Example:
     * findMostFrequent([1, 2, 2, 3, 2]) → 2
     * 
     * @param array Input array
     * @return The most frequent element
     */
    public Integer findMostFrequent(Integer[] array) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Group strings by their length.
     * 
     * Example:
     * groupByLength(["cat", "dog", "bird", "fish"]) → {3: ["cat", "dog"], 4: ["bird", "fish"]}
     * 
     * @param strings Array of strings to group
     * @return Map grouping strings by length
     */
    public Map<Integer, List<String>> groupByLength(String[] strings) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Merge two maps, summing values for duplicate keys.
     * 
     * Example:
     * mergeMaps({"a": 1, "b": 2}, {"b": 3, "c": 4}) → {"a": 1, "b": 5, "c": 4}
     * 
     * @param map1 First map
     * @param map2 Second map
     * @return Merged map with summed values
     */
    public Map<String, Integer> mergeMaps(Map<String, Integer> map1, Map<String, Integer> map2) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Find keys that exist in both maps.
     * 
     * Example:
     * findCommonKeys({"a": 1, "b": 2, "c": 3}, {"b": 4, "c": 5, "d": 6}) → ["b", "c"]
     * 
     * @param map1 First map
     * @param map2 Second map
     * @return List of common keys
     */
    public List<String> findCommonKeys(Map<String, Integer> map1, Map<String, Integer> map2) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Invert a map (swap keys and values).
     * Assumes all values are unique.
     * 
     * Example:
     * invertMap({"a": 1, "b": 2, "c": 3}) → {1: "a", 2: "b", 3: "c"}
     * 
     * @param map Original map
     * @return Inverted map
     */
    public Map<Integer, String> invertMap(Map<String, Integer> map) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Filter map entries based on a condition.
     * Keep entries where value is greater than threshold.
     * 
     * Example:
     * filterByValue({"a": 1, "b": 5, "c": 3}, 2) → {"b": 5, "c": 3}
     * 
     * @param map Original map
     * @param threshold Minimum value threshold
     * @return Filtered map
     */
    public Map<String, Integer> filterByValue(Map<String, Integer> map, Integer threshold) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Calculate word frequency in a text passage.
     * Split by spaces and count occurrences (case-insensitive).
     * 
     * Example:
     * wordFrequency("the cat sat on the mat") → {"the": 2, "cat": 1, "sat": 1, "on": 1, "mat": 1}
     * 
     * @param text Input text
     * @return Map of word frequencies
     */
    public Map<String, Integer> wordFrequency(String text) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Create a map that maintains insertion order.
     * 
     * Example:
     * createLinkedMap(["first", "second", "third"], [1, 2, 3]) → LinkedHashMap preserving order
     * 
     * @param keys Array of keys
     * @param values Array of values
     * @return LinkedHashMap maintaining insertion order
     */
    public LinkedHashMap<String, Integer> createLinkedMap(String[] keys, Integer[] values) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Create a sorted map (TreeMap).
     * 
     * Example:
     * createSortedMap(["zebra", "apple", "banana"], [3, 1, 2]) → TreeMap sorted by keys
     * 
     * @param keys Array of keys
     * @param values Array of values
     * @return TreeMap sorted by natural key ordering
     */
    public TreeMap<String, Integer> createSortedMap(String[] keys, Integer[] values) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Find the key with the maximum value.
     * 
     * Example:
     * findMaxValueKey({"a": 1, "b": 5, "c": 3}) → "b"
     * 
     * @param map Input map
     * @return Key with maximum value, or null if map is empty
     */
    public String findMaxValueKey(Map<String, Integer> map) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Check if two maps are equal in content (same key-value pairs).
     * 
     * Example:
     * mapsEqual({"a": 1, "b": 2}, {"b": 2, "a": 1}) → true
     * 
     * @param map1 First map
     * @param map2 Second map
     * @return true if maps have same key-value pairs
     */
    public boolean mapsEqual(Map<String, Integer> map1, Map<String, Integer> map2) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Create a frequency map from an array.
     * 
     * Example:
     * createFrequencyMap([1, 2, 2, 3, 1, 1]) → {1: 3, 2: 2, 3: 1}
     * 
     * @param array Input array
     * @return Map of element frequencies
     */
    public Map<Integer, Integer> createFrequencyMap(Integer[] array) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Get all keys from the map as a sorted list.
     * 
     * Example:
     * getSortedKeys({"zebra": 1, "apple": 2, "banana": 3}) → ["apple", "banana", "zebra"]
     * 
     * @param map Input map
     * @return Sorted list of keys
     */
    public List<String> getSortedKeys(Map<String, Integer> map) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Demo method to show map operations.
     */
    public static void main(String[] args) {
        MapDictionaryExercises exercises = new MapDictionaryExercises();
        
        System.out.println("Task 5: Maps and Dictionaries - Java");
        System.out.println("====================================");
        System.out.println("Implement the methods to see results!");
        
        try {
            // This will throw UnsupportedOperationException until implemented
            String[] keys = {"a", "b", "c"};
            Integer[] values = {1, 2, 3};
            Map<String, Integer> result = exercises.createMap(keys, values);
            System.out.println("Created map: " + result);
        } catch (UnsupportedOperationException e) {
            System.out.println("Method not implemented: " + e.getMessage());
        }
    }
}

File: python/src/exercises/task05/map_dictionary_exercises.py
"""
Task 5: Maps and Dictionaries - Python
======================================

This module contains exercises for learning Python dictionaries and mapping operations.
Dictionaries are Python's built-in associative data structure, providing efficient
key-value storage with rich functionality and syntax support.

Learning objectives:
- Master dictionary creation, access, and modification
- Use dictionary comprehensions effectively
- Implement frequency counting and grouping algorithms
- Apply advanced dictionary methods and operations
- Understand dictionary iteration patterns
- Learn about collections.defaultdict and Counter
"""

from typing import Dict, List, Optional, Any, Tuple
from collections import defaultdict, Counter


class MapDictionaryExercises:
    """
    Exercises for practicing Python dictionaries and mapping operations.
    
    Each method should be implemented to pass the corresponding unit tests.
    Focus on understanding Python's dictionary capabilities and idiomatic patterns.
    """
    
    def create_dict(self, keys: List[str], values: List[int]) -> Dict[str, int]:
        """
        Create a dictionary from lists of keys and values.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.create_dict(['a', 'b', 'c'], [1, 2, 3])
        {'a': 1, 'b': 2, 'c': 3}
        
        Args:
            keys: List of keys
            values: List of values
            
        Returns:
            Dictionary with paired keys and values
        """
        raise NotImplementedError("Implement create_dict")
    
    def get_value_or_default(self, dictionary: Dict[str, int], key: str, default_value: int) -> int:
        """
        Get a value from the dictionary, returning a default if key doesn't exist.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.get_value_or_default({'a': 1, 'b': 2}, 'a', 0)
        1
        >>> exercises.get_value_or_default({'a': 1, 'b': 2}, 'c', 0)
        0
        
        Args:
            dictionary: The dictionary to search
            key: The key to look up
            default_value: Value to return if key not found
            
        Returns:
            The value for the key, or default value
        """
        raise NotImplementedError("Implement get_value_or_default")
    
    def count_character_frequency(self, text: str) -> Dict[str, int]:
        """
        Count the frequency of each character in a string.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.count_character_frequency('hello')
        {'h': 1, 'e': 1, 'l': 2, 'o': 1}
        
        Args:
            text: Input string
            
        Returns:
            Dictionary of character frequencies
        """
        raise NotImplementedError("Implement count_character_frequency")
    
    def find_most_frequent(self, items: List[int]) -> int:
        """
        Find the most frequent element in a list.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.find_most_frequent([1, 2, 2, 3, 2])
        2
        
        Args:
            items: Input list
            
        Returns:
            The most frequent element
        """
        raise NotImplementedError("Implement find_most_frequent")
    
    def group_by_length(self, strings: List[str]) -> Dict[int, List[str]]:
        """
        Group strings by their length.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.group_by_length(['cat', 'dog', 'bird', 'fish'])
        {3: ['cat', 'dog'], 4: ['bird', 'fish']}
        
        Args:
            strings: List of strings to group
            
        Returns:
            Dictionary grouping strings by length
        """
        raise NotImplementedError("Implement group_by_length")
    
    def merge_dicts(self, dict1: Dict[str, int], dict2: Dict[str, int]) -> Dict[str, int]:
        """
        Merge two dictionaries, summing values for duplicate keys.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.merge_dicts({'a': 1, 'b': 2}, {'b': 3, 'c': 4})
        {'a': 1, 'b': 5, 'c': 4}
        
        Args:
            dict1: First dictionary
            dict2: Second dictionary
            
        Returns:
            Merged dictionary with summed values
        """
        raise NotImplementedError("Implement merge_dicts")
    
    def find_common_keys(self, dict1: Dict[str, int], dict2: Dict[str, int]) -> List[str]:
        """
        Find keys that exist in both dictionaries.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> sorted(exercises.find_common_keys({'a': 1, 'b': 2, 'c': 3}, {'b': 4, 'c': 5, 'd': 6}))
        ['b', 'c']
        
        Args:
            dict1: First dictionary
            dict2: Second dictionary
            
        Returns:
            List of common keys
        """
        raise NotImplementedError("Implement find_common_keys")
    
    def invert_dict(self, dictionary: Dict[str, int]) -> Dict[int, str]:
        """
        Invert a dictionary (swap keys and values).
        Assumes all values are unique.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.invert_dict({'a': 1, 'b': 2, 'c': 3})
        {1: 'a', 2: 'b', 3: 'c'}
        
        Args:
            dictionary: Original dictionary
            
        Returns:
            Inverted dictionary
        """
        raise NotImplementedError("Implement invert_dict")
    
    def filter_by_value(self, dictionary: Dict[str, int], threshold: int) -> Dict[str, int]:
        """
        Filter dictionary entries based on a condition.
        Keep entries where value is greater than threshold.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.filter_by_value({'a': 1, 'b': 5, 'c': 3}, 2)
        {'b': 5, 'c': 3}
        
        Args:
            dictionary: Original dictionary
            threshold: Minimum value threshold
            
        Returns:
            Filtered dictionary
        """
        raise NotImplementedError("Implement filter_by_value")
    
    def word_frequency(self, text: str) -> Dict[str, int]:
        """
        Calculate word frequency in a text passage.
        Split by spaces and count occurrences (case-insensitive).
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.word_frequency('the cat sat on the mat')
        {'the': 2, 'cat': 1, 'sat': 1, 'on': 1, 'mat': 1}
        
        Args:
            text: Input text
            
        Returns:
            Dictionary of word frequencies
        """
        raise NotImplementedError("Implement word_frequency")
    
    def dict_comprehension_example(self, items: List[int]) -> Dict[int, int]:
        """
        Use dictionary comprehension to create squares mapping.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.dict_comprehension_example([1, 2, 3, 4])
        {1: 1, 2: 4, 3: 9, 4: 16}
        
        Args:
            items: List of numbers
            
        Returns:
            Dictionary mapping numbers to their squares
        """
        raise NotImplementedError("Implement dict_comprehension_example")
    
    def find_max_value_key(self, dictionary: Dict[str, int]) -> Optional[str]:
        """
        Find the key with the maximum value.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.find_max_value_key({'a': 1, 'b': 5, 'c': 3})
        'b'
        
        Args:
            dictionary: Input dictionary
            
        Returns:
            Key with maximum value, or None if dictionary is empty
        """
        raise NotImplementedError("Implement find_max_value_key")
    
    def dicts_equal(self, dict1: Dict[str, int], dict2: Dict[str, int]) -> bool:
        """
        Check if two dictionaries are equal in content.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.dicts_equal({'a': 1, 'b': 2}, {'b': 2, 'a': 1})
        True
        
        Args:
            dict1: First dictionary
            dict2: Second dictionary
            
        Returns:
            True if dictionaries have same key-value pairs
        """
        raise NotImplementedError("Implement dicts_equal")
    
    def create_frequency_dict(self, items: List[int]) -> Dict[int, int]:
        """
        Create a frequency dictionary from a list.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.create_frequency_dict([1, 2, 2, 3, 1, 1])
        {1: 3, 2: 2, 3: 1}
        
        Args:
            items: Input list
            
        Returns:
            Dictionary of element frequencies
        """
        raise NotImplementedError("Implement create_frequency_dict")
    
    def get_sorted_keys(self, dictionary: Dict[str, int]) -> List[str]:
        """
        Get all keys from the dictionary as a sorted list.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.get_sorted_keys({'zebra': 1, 'apple': 2, 'banana': 3})
        ['apple', 'banana', 'zebra']
        
        Args:
            dictionary: Input dictionary
            
        Returns:
            Sorted list of keys
        """
        raise NotImplementedError("Implement get_sorted_keys")
    
    def defaultdict_example(self, strings: List[str]) -> Dict[str, List[int]]:
        """
        Use defaultdict to group string indices by first character.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> result = exercises.defaultdict_example(['apple', 'banana', 'cherry', 'apricot'])
        >>> result['a']
        [0, 3]
        >>> result['b']
        [1]
        
        Args:
            strings: List of strings
            
        Returns:
            Dictionary mapping first characters to indices
        """
        raise NotImplementedError("Implement defaultdict_example")
    
    def counter_example(self, text: str) -> Dict[str, int]:
        """
        Use Counter to count character frequencies.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.counter_example('hello')
        {'h': 1, 'e': 1, 'l': 2, 'o': 1}
        
        Args:
            text: Input string
            
        Returns:
            Dictionary of character frequencies using Counter
        """
        raise NotImplementedError("Implement counter_example")
    
    def nested_dict_access(self, nested: Dict[str, Dict[str, int]], path: List[str]) -> Optional[int]:
        """
        Safely access value in nested dictionary using path.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> nested = {'level1': {'level2': {'value': 42}}}
        >>> exercises.nested_dict_access(nested, ['level1', 'level2', 'value'])
        42
        >>> exercises.nested_dict_access(nested, ['level1', 'missing'])
        None
        
        Args:
            nested: Nested dictionary structure
            path: List of keys representing path
            
        Returns:
            Value at path, or None if path doesn't exist
        """
        raise NotImplementedError("Implement nested_dict_access")
    
    def dict_from_tuples(self, tuples: List[Tuple[str, int]]) -> Dict[str, int]:
        """
        Create dictionary from list of tuples.
        
        Example:
        >>> exercises = MapDictionaryExercises()
        >>> exercises.dict_from_tuples([('a', 1), ('b', 2), ('c', 3)])
        {'a': 1, 'b': 2, 'c': 3}
        
        Args:
            tuples: List of (key, value) tuples
            
        Returns:
            Dictionary created from tuples
        """
        raise NotImplementedError("Implement dict_from_tuples")


def main():
    """
    Demo function showing basic dictionary operations.
    """
    exercises = MapDictionaryExercises()
    
    # This will raise NotImplementedError until methods are implemented
    try:
        print("Task 5: Maps and Dictionaries - Python")
        print("======================================")
        print("Implement the methods to see results!")
        
        # Example of what will work once implemented:
        # result = exercises.create_dict(['a', 'b', 'c'], [1, 2, 3])
        # print(f"Created dictionary: {result}")
        
    except NotImplementedError as e:
        print(f"Method not implemented: {e}")


if __name__ == "__main__":
    main()
